### PR TITLE
Auto-detect hex encoded bytes in sha3

### LINKED ISSUE #4081: SHA3 stuck on flawed "binary" encoding
`parity.api.util.sha3` interprets its argument as 'binary' information. This is problematic since there are certain binary blobs that JS cannot encode as a string, which get opaquely coerced to some other string. 

`web.sha3` addresses this by interpreting `0x` prefixed strings as hex-encoded data - other inputs are assumed to be binary-block strings.

It should probably be consistent with `web3.sha3`. Failing that, it should remain consistent with the JSONRPC's conventions and only accept `hex-encoded` data.

### PR DESCRIPTION
**Pull request by tomusdrw**
Closes #4081 

@gavofyork TBH I find previous behaviour more sensible, we have couple of contexts where we explicitly don't want to interpret strings starting with `0x` as bytes (e-mail addresses, solidity function names, etc.).

It would make sense if sha3 accepted ONLY 0x-prefixed, hex-encoded strings, then in most context we would use it as: `sha3(asciiToHex("asdf"))`, right now we kind of mix both approaches.

### CODE REVIEW

File: js/src/api/util/format.js [hunk 1]
```diff
@@ -67,3 +68,7 @@ export function toHex (str) {
 
   return `0x${(str || '').toLowerCase()}`;
 }
+
+export function isHex(str) {
```
Comment (jacogr) [inline js/src/api/util/format.js]: `util/types.js` has an isHex - it is very confusing having the 2 now, not sure what the differences are and what the various uses would be - especially from an external point. In addition, this file is for formatting functions, `util/types` would be for testing.

File: js/src/api/util/sha3.js [hunk 2]
```diff
@@ -28,3 +30,4 @@ export function sha3 (value, options) {
 
   return `0x${hash}`;
 }
+sha3.text = (val) => sha3(val, { encoding: 'raw' });
```
Comment (jacogr) [inline js/src/api/util/sha3.js]: Extra line break required before, no squashing.

File: js/src/api/util/types.js [hunk 3]
```diff
@@ -29,6 +29,10 @@ export function isFunction (test) {
 }
 
 export function isHex (_test) {
+  if (!isString(_test)) {
```
Comment (jacogr) [inline js/src/api/util/types.js]: +1 Was just wondering about this. (Apart from the use-case, makes it more robust overall)

File: js/src/api/util/format.js [hunk 4]
```diff
@@ -15,6 +15,7 @@
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
 import { range } from 'lodash';
+import { isString } from './types';
```
Comment (jacogr) [inline js/src/api/util/format.js]: Unused.

File: js/src/api/util/format.spec.js [hunk 5]
```diff
@@ -14,7 +14,7 @@
 // You should have received a copy of the GNU General Public License
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 
-import { bytesToHex, hexToBytes, hexToAscii, bytesToAscii, asciiToHex } from './format';
+import { bytesToHex, hexToBytes, hexToAscii, bytesToAscii, asciiToHex, isHex } from './format';
```
Comment (jacogr) [inline js/src/api/util/format.spec.js]: Unused.

**Comment by gavofyork** [comment]
> It would make sense if sha3 accepted ONLY 0x-prefixed, hex-encoded strings, then in most context we would use it as: sha3(asciiToHex("asdf")), right now we kind of mix both approaches

This was the original approach, fully consistent with our RPC and I'm happy with it.

That said, I don't see any huge issue with allowing a second `{encoding: 'raw'}` argument to force raw interpretation.

File: js/src/contracts/registry.js [hunk 6]
```diff
@@ -91,7 +91,7 @@ export default class Registry {
 
   lookupAddress (_name) {
     const name = _name.toLowerCase();
-    const sha3 = this._api.util.sha3(name);
+    const sha3 = this._api.util.sha3.text(name);
```
Comment (jacogr) [inline js/src/contracts/registry.js]: Ugly as sin, but ok, I don't have any other suggestions atm ;)

**Comment by gavofyork** [comment]
tests failing

**Comment by gavofyork** [comment]
fwiw i don't think solidity function names are much of an issue since they can't start with `0`.

**Pull request merged**